cmake/module/CheckSourceCompilesAndLinks.cmake

```diff
@@ -12,13 +12,33 @@ macro(check_cxx_source_links source)
   set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
 endmacro()
 
-macro(check_cxx_source_compiles_with_flags flags source)
-  cmake_push_check_state(RESET)
-  set(CMAKE_REQUIRED_FLAGS ${flags})
-  list(JOIN CMAKE_REQUIRED_FLAGS " " CMAKE_REQUIRED_FLAGS)
-  check_cxx_source_compiles("${source}" ${ARGN})
-  cmake_pop_check_state()
-endmacro()
+#[=[
+Check once if C++ source code can be compiled.
+
+Options:
+
+  CXXFLAGS - A list of additional flags to pass to the compiler.
+
+  LDFLAGS - A list of additional flags to pass to the linker.
+
+  LINK_LIBRARIES - A list of libraries to add to the link command.
+
+For historical reasons, among the CMake `CMAKE_REQUIRED_*` variables that influence
+`check_cxx_source_compiles()`, only `CMAKE_REQUIRED_FLAGS` is a string rather than
+a list. Additionally, `target_compile_options()` also expects a list of options.
+
+The `check_cxx_source_compiles_with_flags()` function handles this case and accepts
+`CXXFLAGS` as a list, simplifying the code at the caller site.
+
+]=]
+function(check_cxx_source_compiles_with_flags source result_var)
+  cmake_parse_arguments(PARSE_ARGV 2 _ "" "" "CXXFLAGS;LDFLAGS;LINK_LIBRARIES")
+  list(JOIN __CXXFLAGS " " CMAKE_REQUIRED_FLAGS)
+  set(CMAKE_REQUIRED_LIBRARIES ${__LDFLAGS} ${__LINK_LIBRARIES})
```

Comment:
Why does `CMAKE_REQUIRED_LIBRARIES` get `LDFLAGS` rather than setting `LINK_OPTIONS`?

Reply:
> I think a static lib would work fine as an option though?

Correct. This is documented by CMake.

Reply:
I meant as `LINK_OPTIONS`. I'd rather have LDFLAGS forward to that unless there's a reason not to.

Reply:
Okay, I've checked CMake docs again.

`CMAKE_REQUIRED_LINK_OPTIONS` is forwarded to `target_link_options()`, which is [designed](https://cmake.org/cmake/help/latest/command/target_link_options.html) "to add any link options".

I'll update the PR shortly.

Reply:
> I meant as `LINK_OPTIONS`. I'd rather have LDFLAGS forward to that unless there's a reason not to.

In that case a static library (e.g., `/usr/lib/libFuzzingEngine.a`) would be placed _before_ object files being linked, and the linker might fail.

Reply:
Yeah, I really don't think we should be worrying about their busted flags.

For OSS-Fuzz, we can change it to pass:
`-DSANITIZER_LDFLAGS="-Wl,$LIB_FUZZING_ENGINE"`. That should fix the ordering problem if it's an absolute file while still working with "-lFuzzingEngine" as well.

Reply:
Blah, ok, sorry. I've been messing around with this. Final answer:

- Let's add a `SANITIZER_LIBS` var
- add BOTH `LINK_LIBRARIES` and `LDFLAGS` to check_cxx_source_compiles_with_flags
- pass them both for the sanitizers
- update oss-fuzz to use `SANITIZER_LIBS` instead.

That way our stuff is well-defined and makes sense and we're not coding around oss-fuzz, but we still expose the option we need there.

Reply:
cc the OSS-Fuzz crowd: @maflcko @dergoegge @fanquake @achow101

Reply:
@fanquake Are you ok if we break OSS-Fuzz here with a proper fix (to avoid future technical debt), and PR the one-liner to fix it there post-merge?

Reply:
Yes